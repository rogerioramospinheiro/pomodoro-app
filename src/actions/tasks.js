export const startTask = ({title, description, start_date, activity_time, rest_time} = {}) => ({
    type: 'START_TASK',
    task: {
        title,
        description,
        start_date,
        activity_time,
        rest_time
    }
});

export const cancelTask = () => ({
    type: 'CANCEL_TASK'
});

export const finishTask = ({end_date} = {}) => ({
    type: 'FINISH_TASK',
    task: {
        end_date
    }
});