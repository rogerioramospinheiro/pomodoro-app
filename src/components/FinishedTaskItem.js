import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';

const FinishedTaskItem = ( {id, title, description, start_date, end_date} ) => (
    <Link className="finished-tasks__link" to={`/edit/${id}`}>
        <div className="finished-tasks__item">
            <div>{ title }</div>
            <div>{ description }</div>
            <div>
                <span>{ moment(start_date).format('hh:mm:ss A') }</span>
                <span>-</span>
                <span>{ moment(end_date).format('hh:mm:ss A') }</span>
            </div>
        </div>
    </Link>
);

export default FinishedTaskItem;