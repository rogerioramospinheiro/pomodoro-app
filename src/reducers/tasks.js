import uuidv1 from 'uuid/v1';

const tasksReducerDefaultState = {
    state_name: 'IDLE',
    current_task: {
    },
    finished_tasks: []
};

const transitionStartToCounting = (state, action) => {
    if (state.state_name === 'IDLE') {
        return {
            state_name: 'COUNTING',
            current_task: {
                title: action.task.title,
                description: action.task.description,
                start_date: action.task.start_date,
            },
            finished_tasks: [...state.finished_tasks]
        };
    } else {
        return state;
    }
};

const transitionCancelToIdle = (state) => {
    if (state.state_name !== 'IDLE') {
        return {
            state_name: 'IDLE',
            current_task: tasksReducerDefaultState.current_task,
            finished_tasks: [...state.finished_tasks]
        }
    } else {
        return state;
    }
};

const transitionFinishToIdle = (state, action) => {
    if (state.state_name === 'COUNTING') {
        return {
            state_name: 'RESTING',
            current_task: {
                title: state.current_task.title,
                description: state.current_task.description,
                start_date: state.current_task.start_date,
            },
            finished_tasks: [
                ...state.finished_tasks
            ]
        }
    } else if (state.state_name === 'RESTING') {
        const finished_task = {
            id: uuidv1(),
            title: state.current_task.title,
            description: state.current_task.description,
            start_date: state.current_task.start_date,
            end_date: action.task.end_date
        };
        return {
            state_name: 'IDLE',
            current_task: tasksReducerDefaultState.current_task,
            finished_tasks: [
                finished_task,
                ...state.finished_tasks
            ]
        }
    } else {
        return state;
    }
}

const completeEditFinishedTask = (state, action) => {
    const finished_tasks = state.finished_tasks.map( (task) => {
        if (task.id === action.id) {
            return {
                ...task,
                ...action.updates
            };
        } else {
            return task;
        }
    });
    return {
        ...state,
        finished_tasks
    };
}

export default (state = tasksReducerDefaultState, action) => {
    switch(action.type) {
        case 'START_TASK':
            return transitionStartToCounting(state, action);
        case 'CANCEL_TASK':
            return transitionCancelToIdle(state);
        case 'FINISH_TASK':
            return transitionFinishToIdle(state, action);
        case 'EDIT_FINISHED_TASK':
            return completeEditFinishedTask(state, action);
        default:
            return state;
    }
};