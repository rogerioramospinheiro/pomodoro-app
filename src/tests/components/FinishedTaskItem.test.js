import React from 'react';
import { shallow } from 'enzyme';
import { createFinishedTasksState } from '../fixtures/tasks';
import FinishedTaskItem from '../../components/FinishedTaskItem';

test('Should render FinishedTaskItem', () => {
    // Given
    const state = createFinishedTasksState();
    const finishedTask = state.finished_tasks[0];

    // When
    const wrapper = shallow(<FinishedTaskItem {...finishedTask} />);

    // Then
    expect(wrapper).toMatchSnapshot();
});
