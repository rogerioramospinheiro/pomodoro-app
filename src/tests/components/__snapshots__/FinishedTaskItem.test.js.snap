// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render FinishedTaskItem 1`] = `
<Link
  className="finished-tasks__link"
  to="/edit/c638040c-3917-11ea-a137-2e728ce88125"
>
  <div
    className="finished-tasks__item"
  >
    <div>
      TASK-0001
    </div>
    <div>
      Job to be finished soon
    </div>
    <div>
      <span>
        07:00:10 PM
      </span>
      <span>
        -
      </span>
      <span>
        07:00:30 PM
      </span>
    </div>
  </div>
</Link>
`;
